### ISSUE TITLE
embedding visualization example

### ISSUE THREAD

**Issue by hkhatod** [tensorflow/tensorboard#805]
Can someone please have end to end embedding visualization example. The documentation is fragmented  I am not even sure if I have all the required code implemented correctly. Please dont close the issue for lack of details....I am not even able to get started with embedding viz...I dont have meaning full code to post here.

**Comment by vitalyli** [comment]
I will post some example later in the day.

**Comment by hkhatod** [comment]
Cant wait. Thanks @vitalyli vitalyli

**Comment by hkhatod** [comment]
This is whats happening. I see a wheel spinning with parsing metadata message. Any clues as to what might be wrong 

![screenshot from 2017-12-10 19-34-22](https://user-images.githubusercontent.com/23573076/33812255-5b838f0c-dde1-11e7-9d2b-83582ae6ab79.png)

**Comment by hkhatod** [comment]
Also, note it says Point 0 and Dimensions 0

**Comment by vitalyli** [comment]
Here is example (note: you need to parse embedding file and load those into TF graph variable, then metadata file to add metadata such as name associated with id)

# -*- coding: utf-8 -*-
import os
import tensorflow as tf
import numpy as np
import argparse
import datetime
import codecs

from tensorflow.contrib.tensorboard.plugins import projector

import emb_dict

if __name__ == "__main__":

    run_dt = datetime.datetime.now().strftime("%Y%m%d%H%M%S")

    fs = './out_dir'
    log_dir = fs + '/viz/emb_viz_%s' % run_dt

    emb_file_path = '/feature_store/embeddings/vectors_20171127005623.txt'

    parser = argparse.ArgumentParser()

    parser.add_argument(
            "--emb_file",
            type=str,
            default=emb_file_path,
            help="emb_file"
    )
    parser.add_argument(
            "--emb_dim",
            type=int,
            default=100,
            help="emb_dim"
    )
    parser.add_argument(
            "--emb_type",
            type=str,
            default="glove",
            help="emb_type"
    )

    parser.add_argument(
            "--log_dir",
            type=str,
            default=log_dir,
            help="log_dir"
    )

    parser.add_argument(
            "--name_file",
            type=str,
            default=None,
            help="name_file"
    )

    FLAGS, unparsed = parser.parse_known_args()

    print('name_file', FLAGS.name_file)
    name_dict = {}
    if FLAGS.name_file:
        with codecs.open(FLAGS.name_file, 'r', encoding='utf8') as n_file:
            for line in n_file:
                splitt = line.lower().strip().split('\t')
                app_adam = splitt[0].strip()
                app_name = "none"
                if len(splitt) > 1:
                    app_name = splitt[1].strip()

                name_dict[app_adam] = app_name

    FLAGS.log_dir = FLAGS.log_dir + '/' + run_dt

    if not os.path.exists(FLAGS.log_dir):
        os.makedirs(FLAGS.log_dir)

    print("Output dir:",FLAGS.log_dir)

    embDic = emb_dict.EmdDict(FLAGS.emb_file, dim=FLAGS.emb_dim, tpe=FLAGS.emb_type)
    embDic.load_data()

    tf.reset_default_graph()
    sess = tf.InteractiveSession()
    X = tf.Variable([0.0], name='embedding')
    place = tf.placeholder(tf.float32, shape=[embDic.vocab_size, embDic.dim])
    set_x = tf.assign(X, place, validate_shape=False)

    sess.run(tf.global_variables_initializer())
    sess.run(set_x, feed_dict={place: embDic.embeddings})

    # write labels
    with codecs.open(FLAGS.log_dir + '/metadata.tsv', 'w', encoding='utf8') as f:
        if len(name_dict) > 0:
            f.write("id\tname\n".encode('utf-8'))
            for word in embDic.vocab:
                found_name = name_dict.get(word, "unk")
                line_out = "%s\t%s\n".encode('utf-8') % (word,found_name)
                f.write(line_out)
        else:
            for word in embDic.vocab:
                line_out = "%s\n" % word
                f.write(line_out.encode('utf-8'))

    # create a TensorFlow summary writer
    summary_writer = tf.summary.FileWriter(FLAGS.log_dir + '/' + 'emb_viz_%s.log' % run_dt, sess.graph)
    config = projector.ProjectorConfig()
    embedding_conf = config.embeddings.add()
    embedding_conf.tensor_name = 'embedding:0'
    embedding_conf.metadata_path = os.path.join(FLAGS.log_dir, 'metadata.tsv')
    projector.visualize_embeddings(summary_writer, config)

    saver = tf.train.Saver()
    saver.save(sess, os.path.join(FLAGS.log_dir, "model.ckpt"))

**Comment by hkhatod** [comment]
Thanks. But still confused as hell.

Here is my code snippet. I am wanting to visualize embedded_chars.  But nothing loads when I use that as my embedding_conf.tensor_name ". On the other hand I can visualize emb_var but it doenst match my metadat.tsv....Please help
`
    class TextCNNRNN(object):
	def __init__(self, embedding_mat, non_static, hidden_unit, sequence_length, max_pool_size,
	             num_classes, embedding_size, filter_sizes, num_filters, l2_reg_lambda=0.0):

		self.input_x = tf.placeholder(tf.int32, [None, sequence_length], name='input_x')
		self.input_y = tf.placeholder(tf.float32, [None, num_classes], name='input_y')
		self.dropout_keep_prob = tf.placeholder(tf.float32, name='dropout_keep_prob')
		self.batch_size = tf.placeholder(tf.int32, [])
		self.pad = tf.placeholder(tf.float32, [None, 1, embedding_size, 1], name='pad')
		self.real_len = tf.placeholder(tf.int32, [None], name='real_len')

		l2_loss = tf.constant(0.0)

		with tf.device('/cpu:0'), tf.name_scope('embedding'):
			self.emb_var = tf.Variable(embedding_mat, name='emb_var')
			# if not non_static:
			# 	self.emb_var = tf.constant(embedding_mat, name='enb_var')
			# else:
			# 	self.emb_var = tf.Variable(embedding_mat, name='emb_var')
			self.embedded_chars = tf.nn.embedding_lookup(self.emb_var, self.input_x)
			self.emb = tf.expand_dims(self.embedded_chars, -1)
			
			`

**Comment by hkhatod** [comment]
my projecter code:
`	config = projector.ProjectorConfig()
			config.model_checkpoint_path = checkpoint_prefix + str(best_at_step) +'.ckpt'
			embedding = config.embeddings.add()
			embedding.tensor_name = cnn_rnn.emb_var.name
			
			#embedding.metadata_path = 'metadata.tsv'
			emb_writer = tf.summary.FileWriter(metadata_path,sess.graph)
			projector.visualize_embeddings(emb_writer, config)`

**Comment by vitalyli** [comment]
I can't tell for sure, for projector basically every variable is an embedding vector. You need to make sure that variable name is exactly how you tell projector about it. Variable scope as well as device changes that name. Then obviously project has to get ids as vector of same size as variable.

**Comment by hkhatod** [comment]
I think the problem is embedded_chars  is not being saved. How do I make sure its getting saved  when I call the saver.

**Comment by hkhatod** [comment]
OK. I think what I am looking for doesnt exist and may fall in the category of feature request ? Dont know. Please help.


I am working on a text classification problem. I have 3M+ rows which need to be classified into 20 categories.

Following are two code snippets from my entire code:

This is the code where my tf variables are defined.:
```

class TextCNNRNN(object):
    def __init__(self, embedding_mat, non_static, hidden_unit, sequence_length, max_pool_size,
                 num_classes, embedding_size, filter_sizes, num_filters, l2_reg_lambda=0.0):

        self.input_x = tf.placeholder(tf.int32, [None, sequence_length], name='input_x')
        self.input_y = tf.placeholder(tf.float32, [None, num_classes], name='input_y')
        self.dropout_keep_prob = tf.placeholder(tf.float32, name='dropout_keep_prob')
        self.batch_size = tf.placeholder(tf.int32, [])
        self.pad = tf.placeholder(tf.float32, [None, 1, embedding_size, 1], name='pad')
        self.real_len = tf.placeholder(tf.int32, [None], name='real_len')

        l2_loss = tf.constant(0.0)

        with tf.device('/cpu:0'), tf.name_scope('embedding'):
            self.emb_var = tf.Variable(embedding_mat, name='emb_var')
            # if not non_static:
            #   self.emb_var = tf.constant(embedding_mat, name='enb_var')
            # else:
            #   self.emb_var = tf.Variable(embedding_mat, name='emb_var')
            self.embedded_chars = tf.nn.embedding_lookup(self.emb_var, self.input_x)
            self.emb = tf.expand_dims(self.embedded_chars, -1)
```
The tensor I would like to visualize is embedded_chars.

And this is the code where I am given input to the projector api:

```
config = projector.ProjectorConfig()
config.model_checkpoint_path = checkpoint_prefix + str(best_at_step) +'.ckpt'
embedding = config.embeddings.add()
embedding.tensor_name = cnn_rnn.embedded_chars.name

#embedding.metadata_path = 'metadata.tsv'
emb_writer = tf.summary.FileWriter(metadata_path,sess.graph)
projector.visualize_embeddings(emb_writer, config)
```
seems like this functionality is currently not supported. For some reason embedded_chars tensor is just note saved and the project does not read it.

**Comment by jart** [comment]
Thank you for helping our friend @vitalyli. I'll also point out https://www.tensorflow.org/programmers_guide/embedding. We're always open to contributions adding additional README files, and other pull requests. I'm also sure the folks Googling will see this issue in the future, which can be of assistance.

**Comment by jart** [comment]
I would also like to encourage folks in the community to make YouTube videos too.

For example, if you have a Macbook, there's Quicktime built in for free, which can record a screencast where you talk your way through how to use various things.

If anyone puts a lot of effort into creating things like that, then send us a pull request adding it to our README. Maybe we could start a bullet point section with these sorts of tutorial videos. Plus the PR itself gives additional exposure to the video.

**Comment by hkhatod** [comment]
Can someone answer my question before closing though. Is this a feature request or is this something I should be able to do with current tools ?

**Comment by vitalyli** [comment]
Few things I can see:
0. This doesn't load embeddings into variable:
self.emb_var = tf.Variable(embedding_mat, name='emb_var')

->you need tf.constant_initializer or call an op to assign embeddings to a var.
for example:

            embeddingsW = tf.get_variable(name="embeddingsW", shape=[embDic.vocab_size, embDic.dim],
                                          initializer=tf.constant_initializer(np.asmatrix(embDic.embeddings)),
                                          dtype=tf.float32, trainable=False)

1. emb_var variable should be trainable=False otherwise Optimizer will be modifying it, which you don't want.

2. Properly loaded variable would appear under embedding tab by default

3. This is where you supply metadata, for projector to display and allow to search by id or label (if given); you have it commented out for some reason.  #embedding.metadata_path = 'metadata.tsv'

4. what does this line do?
tf.expand_dims(self.embedded_chars, -1)

-> this function needs axis to expand; 
for example: axis=3
